subprojects/core/src/main/java/org/gradle/initialization/DefaultSettingsLoader.java

```diff
@@ -35,45 +38,81 @@
 import java.util.List;
 
 /**
- * Handles locating and processing setting.gradle files.  Also deals with the buildSrc module, since that modules is
+ * Handles locating and processing setting.gradle files.  Also deals with the buildSrc module, since that module is
  * found after settings is located, but needs to be built before settings is processed.
  */
 public class DefaultSettingsLoader implements SettingsLoader {
     public static final String BUILD_SRC_PROJECT_PATH = ":" + SettingsInternal.BUILD_SRC;
+
     private final SettingsProcessor settingsProcessor;
     private final BuildLayoutFactory buildLayoutFactory;
     private final List<BuiltInCommand> builtInCommands;
+    private final Logger logger;
 
     public DefaultSettingsLoader(
         SettingsProcessor settingsProcessor,
         BuildLayoutFactory buildLayoutFactory,
         List<BuiltInCommand> builtInCommands
+    ) {
+        this(settingsProcessor, buildLayoutFactory, builtInCommands, Logging.getLogger(DefaultSettingsLoader.class));
+    }
+
+    @VisibleForTesting
+    /* package */ DefaultSettingsLoader(
+        SettingsProcessor settingsProcessor,
+        BuildLayoutFactory buildLayoutFactory,
+        List<BuiltInCommand> builtInCommands,
+        Logger logger
     ) {
         this.settingsProcessor = settingsProcessor;
         this.buildLayoutFactory = buildLayoutFactory;
         this.builtInCommands = builtInCommands;
+        this.logger = logger;
     }
 
     @Override
     public SettingsState findAndLoadSettings(GradleInternal gradle) {
         StartParameter startParameter = gradle.getStartParameter();
-
         SettingsLocation settingsLocation = buildLayoutFactory.getLayoutFor(new BuildLayoutConfiguration(startParameter));
 
-        SettingsState state = findSettingsAndLoadIfAppropriate(gradle, startParameter, settingsLocation, gradle.getClassLoaderScope());
-        SettingsInternal settings = state.getSettings();
-        ProjectSpec spec = ProjectSpecs.forStartParameter(startParameter, settings);
-        if (useEmptySettings(spec, settings, startParameter)) {
-            // Discard the loaded settings and replace with an empty one
-            state.close();
-            state = createEmptySettings(gradle, startParameter, settings.getClassLoaderScope());
-            settings = state.getSettings();
+        SettingsState state;
+        ProjectSpec spec;
+        if (shouldSkipLoadingBuildDefinition(startParameter)) {
```

Comment:
Do we have a test that checks what happens when there's a settings file and we run init from that directory?  What about along with other tasks?


I think another side effect of this change is if you were in a directory with a settings file that had an error in it, I'm not sure you could run `gradle init` at all before?

Reply:
So I added a couple new tests for this and ran them against `master`.  Here's what has changed:

- `fails when initializing in a directory that contains a working settings file` - No change, still fails with existing files in project dir error.
- `fails when initializing in a directory that contains an invalid settings file` - Previously failed with error evaluating settings file, now fails with existing files in project dir error.
- `fails when initializing plus help in a directory that contains a working settings file` - No change, still fails with existing files in project dir error.

So I don't think there's an issue here, just moving from one type of failure to another in the invalid settings case.